go/vt/schemadiff/column.go

```diff
@@ -71,11 +71,69 @@ func NewModifyColumnDiffByDefinition(definition *sqlparser.ColumnDefinition) *Mo
 }
 
 type ColumnDefinitionEntity struct {
-	columnDefinition *sqlparser.ColumnDefinition
+	columnDefinition    *sqlparser.ColumnDefinition
+	tableCharsetCollate *charsetCollate
+	Env                 *Environment
 }
 
-func NewColumnDefinitionEntity(c *sqlparser.ColumnDefinition) *ColumnDefinitionEntity {
-	return &ColumnDefinitionEntity{columnDefinition: c}
+func NewColumnDefinitionEntity(env *Environment, c *sqlparser.ColumnDefinition, tableCharsetCollate *charsetCollate) *ColumnDefinitionEntity {
+	return &ColumnDefinitionEntity{
+		columnDefinition:    c,
+		tableCharsetCollate: tableCharsetCollate,
+		Env:                 env,
+	}
+}
+
+func (c *ColumnDefinitionEntity) Clone() *ColumnDefinitionEntity {
+	clone := &ColumnDefinitionEntity{
+		columnDefinition:    sqlparser.Clone(c.columnDefinition),
+		tableCharsetCollate: c.tableCharsetCollate,
+		Env:                 c.Env,
+	}
+	return clone
+}
+
+// SetExplicitCharsetCollate enriches this column definition with collation and charset. Those may be
+// already present, or perhaps just one of them is present (in which case we use the one to populate the other),
+// or both might be missing, in which case we use the table's charset/collation.
+func (c *ColumnDefinitionEntity) SetExplicitCharsetCollate() error {
+	if !c.IsTextual() {
+		return nil
+	}
+	// We will now denormalize the columns charset & collate as needed (if empty, populate from table.)
+	// Normalizing _this_ column definition:
+	if c.columnDefinition.Type.Charset.Name != "" && c.columnDefinition.Type.Options.Collate == "" {
+		// Charset defined without collation. Assign the default collation for that charset.
+		collation := c.Env.CollationEnv().DefaultCollationForCharset(c.columnDefinition.Type.Charset.Name)
+		if collation == collations.Unknown {
+			return &UnknownColumnCharsetCollationError{Column: c.columnDefinition.Name.String(), Charset: c.tableCharsetCollate.charset}
+		}
+		c.columnDefinition.Type.Options.Collate = c.Env.CollationEnv().LookupName(collation)
+	}
+	if c.columnDefinition.Type.Charset.Name == "" && c.columnDefinition.Type.Options.Collate != "" {
+		// Column has explicit collation but no charset. We can infer the charset from the collation.
+		collationID := c.Env.CollationEnv().LookupByName(c.columnDefinition.Type.Options.Collate)
+		charset := c.Env.CollationEnv().LookupCharsetName(collationID)
+		if charset == "" {
+			return &UnknownColumnCollationCharsetError{Column: c.columnDefinition.Name.String(), Collation: c.columnDefinition.Type.Options.Collate}
+		}
+		c.columnDefinition.Type.Charset.Name = charset
+	}
+	if c.columnDefinition.Type.Charset.Name == "" {
+		// Still nothing? Assign the table's charset/collation.
+		c.columnDefinition.Type.Charset.Name = c.tableCharsetCollate.charset
+		if c.columnDefinition.Type.Options.Collate == "" {
+			c.columnDefinition.Type.Options.Collate = c.tableCharsetCollate.collate
+		}
```

Comment:
Why do this when we blindly set `c.columnDefinition.Type.Options.Collate = c.tableCharsetCollate.collate` in the subsequent IF condition?

Reply:
Good catch! Removed.